cborg/tests/Tests/FlatTerm.hs

```diff
@@ -0,0 +1,121 @@
+termToInValidTermTokens (TStrings ss)  = TkStringBegin : (TkString . T.pack <$> ss) ++ [TkBytesBegin]
+termToInValidTermTokens (TArray ts)    =
+  TkListLen (fromIntegral $ length ts) : (mconcat $ termToInValidTermTokens <$> ts) ++ [TkNull]
+termToInValidTermTokens (TArrayI ts)   =
+  TkListBegin : TkListBegin : (mconcat $ termToInValidTermTokens <$> ts) ++ [TkBreak]
+termToInValidTermTokens (TMap kvs)     =
+  let kvList = mconcat $ (\ (a, _) -> [a]) <$> kvs
+  in TkMapLen (fromIntegral $ length kvList) : (mconcat $ termToInValidTermTokens <$> kvList) ++ [TkNull]
+termToInValidTermTokens (TMapI kvs)    =
+  let kvList = mconcat $ (\ (a, _) -> [a]) <$> kvs
+  in TkMapBegin : TkStringBegin : (mconcat $ termToInValidTermTokens <$> kvList) ++ [TkBreak]
+termToInValidTermTokens x = TkStringBegin : TkNull : TkBreak : termToValidTermTokens x
+
+
+-- Convert Term to List of Valid TermTokens
+termToValidTermTokens :: Term -> [TermToken]
+termToValidTermTokens (TUInt u )     = [ TkInt . fromIntegral $ fromUInt u ]
+termToValidTermTokens (TNInt u )     = [ TkInt . fromIntegral $ fromUInt u ]
+termToValidTermTokens (TBigInt i)    = [ TkInteger i ]
+termToValidTermTokens (TBytes w)     = [ TkBytes (BS.pack w) ]
+termToValidTermTokens (TBytess ws)   = TkBytesBegin : ((TkBytes . BS.pack) <$> ws) ++ [TkBreak]
+termToValidTermTokens (TString s)    = [ TkString (T.pack s) ]
+termToValidTermTokens (TStrings ss)  = TkStringBegin : (TkString . T.pack <$> ss) ++ [TkBreak]
+termToValidTermTokens (TArray ts)    =
+  TkListLen (fromIntegral $ length ts) : (mconcat $ termToValidTermTokens <$> ts)
+termToValidTermTokens (TArrayI ts)   = TkListBegin : (mconcat $ termToValidTermTokens <$> ts) ++ [TkBreak]
+termToValidTermTokens (TMap kvs)     =
+  let kvList = mconcat $ (\ (a, b) -> [a,b]) <$> kvs
+  in TkMapLen (fromIntegral $ length kvs) : (mconcat $ termToValidTermTokens <$> kvList)
+termToValidTermTokens (TMapI kvs)    =
+  let kvList = mconcat $ (\ (a, b) -> [a,b]) <$> kvs
+  in TkMapBegin : (mconcat $ termToValidTermTokens <$> kvList) ++ [TkBreak]
+termToValidTermTokens (TTagged ui t) = TkTag (fromUInt ui) : termToValidTermTokens t
+termToValidTermTokens TTrue          = [ TkBool True ]
+termToValidTermTokens TFalse         = [ TkBool False ]
+termToValidTermTokens TNull          = [ TkNull ]
+termToValidTermTokens TUndef         = [ TkNull ]
+termToValidTermTokens (TSimple s)    = [ TkSimple (fromSimple s) ]
+termToValidTermTokens (TFloat16 hs)  = [ TkFloat16 (fromHalf . getHalfSpecials $ hs) ]
+termToValidTermTokens (TFloat32 fs)  = [ TkFloat32 (getFloatSpecials fs) ]
+termToValidTermTokens (TFloat64 ds)  = [ TkFloat64 (getDoubleSpecials ds) ]
+
+prop_validateValidFlat :: ValidFlat -> Property
+prop_validateValidFlat (ValidFlat ts) = property $ validFlatTerm ts
+
+prop_validateInValidFlat :: InValidFlat -> Property
+prop_validateInValidFlat (InValidFlat ts) = property $ validFlatTerm ts == False
```

Comment:
This is a very add-hoc property. The generator for `InValidFlat` is just using various but fixed ways of making an invalid flat term.

I'm not happy with adding this property. It looks like it's testing something that it's really not testing well.

An alternative approach would be to naively generate arbitrary flat terms and then check that the reference decoder (ie from `Tests.Reference.Implementation`) agrees with `validFlatTerm`. By agree, I mean they both agree it's valid, or both agree it's invalid. Of course most arbitrarily generated flat terms will be invalid.

Reply:
@dcoutts Yes I agree you are right, the generation of `InvalidFlat` is ad hoc in the sense that it is not covering all possible ways of invalidity but a few fixed ones.

Alternative approach looks fine to me btw do you think that testing arbitrarily generated FlatTerms along with the current setup will be better or its of no use?

Reply:
I think the currently proposed change here is misleading, and not of a lot of use. I'd accept the PR without the second patch. Or I'd accept the second patch without the bogus property.

Reply:
Changed the property as per your suggestions!